### PR TITLE
cp2k: fix builds on macOS, workaround reported issue with __contains__

### PR DESCRIPTION
**Pull request by dev-zero**
- netlib-scalapack: fix build on macOS (and possibly others)
- intel-oneapi-mkl: fix lib discovery on macOS for external lib
- cp2k: fix building on non-GNU systems, workaround #35597

### CODE REVIEW

**Comment by dev-zero** [comment]
for cp2k and netlib-scalapack @alalazo or @haampie ?

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 1]
```diff
@@ -153,7 +154,10 @@ def _find_mkl_libs(self, shared):
         if "+cluster" in self.spec:
             libs.append(self._xlp64_lib("libmkl_blacs_intelmpi"))
 
-        return find_libraries(libs, self.component_prefix.lib.intel64, shared=shared)
+        lib_path = join_path(self.component_prefix, "lib", "intel64")
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py]:
```suggestion
        lib_path = self.component_prefix.lib.intel64
```

**Review by rscohn2 — DISMISSED** [review]
Looks good.

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 2]
```diff
@@ -274,31 +275,14 @@ def archive_files(self):
...
-            fftw = spec["intel-parallel-studio"]
-            fftw_header_dir = "<NOTFOUND>"
-            for incdir in [join_path(f, "fftw") for f in fftw.headers.directories]:
-                if os.path.exists(incdir):
-                    fftw_header_dir = incdir
-                    break
-        elif "^cray-fftw" in spec:
-            fftw = spec["cray-fftw"]
-            fftw_header_dir = fftw.headers.directories[0]
+        fftw = spec["fftw-api"]
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py:277]: Here we remove the query-like part of the subscript, for instance:
```python
fftw = spec["fftw:openmp" if "+openmp" in spec else "fftw"]
```
Is it fine to not pass the `:openmp` part when later we use `fftw.libs`?

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 3]
```diff
@@ -401,11 +385,12 @@ def edit(self, spec, prefix):
         ldflags.append((lapack + blas).search_flags)
         libs.extend([str(x) for x in (fftw.libs, lapack, blas)])
 
-        if any(
-            p in spec for p in ("^intel-mkl", "^intel-parallel-studio+mkl", "^intel-oneapi-mkl")
-        ):
+        if sys.platform == "darwin":
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]:
```suggestion
        if self.spec.satisfies("platform=darwin"):
```

**Review by alalazo — COMMENTED** [review]
Basically LGTM. One question and one minor suggestion

File: var/spack/repos/builtin/packages/cp2k/package.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import copy
 import os
 import os.path
+import sys
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/cp2k/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/cp2k/package.py (hunk 2 above)
Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: I hoped you would know :-)
To be honest I was surprised that the `"^provider" in spec` stopped working, but things like `spec["virtual"].libs` seem to be working now. Based on that I assumed that there will only be one provider for "fftw-api" and that will be the right one (as selected by the depend spec before).

Comment (alalazo) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: > To be honest I was surprised that the "^provider" in spec stopped working,

Wait, this should be filed as a bug :thinking: I thought the changes were a mere preference of style by you as the maintainer of the package. Where is it mentioned in this PR that `^provider` is not working?

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: in "fix provider-specific flags"

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: On the other hand, this style is also more correct as it covers using intel-oneapi-mkl for blas/lapack but netlib-scalapack for scalapack.

Comment (alalazo) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: Reporting from the conversation in Slack:
```python
Spack version 0.20.0.dev0
Python 3.8.10, Linux x86_64
>>> from spack.spec import Spec
>>> s = Spec("cp2k")
>>> c = s.concretized()
>>> "openmpi" in c
True
>>> "^openmpi" in c
True
>>> "mpich" in c
False
>>> "^mpich" in c
False
>>> 
```
It seems I cannot reproduce the error (also, not to be pedantic, but it's hard to understand for me that "fix for provider flags" means that the spec syntax is broken for you). Do you have a reproducer for the issue, to investigate it better?

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: I have a `spack-build-out.txt` from a previous attempt:

```
==> cp2k: Executing phase: 'edit'
==> [2023-02-19-21:57:03.468919] '/usr/local/bin/gcc-12' '-dumpversion'
==> [2023-02-19-21:57:05.345084] '/usr/local/bin/pkg-config' '--cflags' 'libint2'
==> [2023-02-19-21:57:05.359064] '/usr/local/bin/pkg-config' '--libs' 'libint2'
==> [2023-02-19-21:57:05.364499] '/usr/local/bin/pkg-config' '--cflags' 'libxcf03'
==> [2023-02-19-21:57:05.371111] '/usr/local/bin/pkg-config' '--libs' 'libxcf03' 'libxc'
==> cp2k: Executing phase: 'build'
==> [2023-02-19-21:57:05.393194] 'make' '-j1' 'ARCH=darwin-ventura-skylake-gcc' 'VERSION=psmp'
Discovering programs ...
Removing stale archives for psmp ...
Resolving dependencies for psmp ...
/Library/Developer/CommandLineTools/usr/bin/make -C /var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/dbcsr -f /var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/build_dbcsr/Makefile \
           ARCHFILE=/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/arch/darwin-ventura-skylake-gcc.psmp \
           LIBDIR=/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/lib/darwin-ventura-skylake-gcc/psmp/exts/dbcsr \
           OBJDIR=/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/obj/darwin-ventura-skylake-gcc/psmp/exts/dbcsr \
           USE_ACCEL="" \
           ACC="" \
           ACCFLAGS=""
Removing stale archives ...
Resolving dependencies ...
/usr/bin/env python3 /private/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/dbcsr/tools/build_utils/fypp/bin/fypp -n /private/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/dbcsr/src/base/dbcsr_kinds.F dbcsr_kinds.F90
/usr/local/bin/mpif90 -c -O2 -funroll-loops -ftree-vectorize -ffree-form -ffree-line-length-none -ggdb -fopenmp -I/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/libint-2.6.0-zhcvkhf3zzyociu7ep22oxzqak2koiws/include -I/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/libint-2.6.0-zhcvkhf3zzyociu7ep22oxzqak2koiws/include/libint2 -I/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/boost-1.80.0-da34dhsijvwcq5l6la42qc2gnlwig4t4/include -I/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/libxc-6.1.0-ipsgzeujk2uhcamoeglwxgdw62dyc3fb/include -D__FFTW3 -I/opt/intel/oneapi/mkl/2023.0.0/include/fftw -D__parallel -D__SCALAPACK -D__LIBINT -D__LIBXC -D__SHORT_FILE__="\"dbcsr_kinds.F\"" -I'/private/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/dbcsr/src/base/' -I'/private/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exts/dbcsr/src' dbcsr_kinds.F90
...
/usr/local/bin/mpif90 -fopenmp -L/opt/intel/oneapi/mkl/2023.0.0/lib -L/opt/intel/oneapi/mkl/2023.0.0/lib -L/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/netlib-scalapack-2.2.0-ymqt25ppsjrbpxn2pg34fbwvrkg4osml/lib   -L/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/lib/darwin-ventura-skylake-gcc/psmp -o /var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/exe/darwin-ventura-skylake-gcc/grid_miniapp.psmp grid_miniapp.o  -lcp2kgrid -lcp2kgridcpu -lcp2kgridref -lcp2kgridcommon -lcp2ksubsys -lcp2kfm -lcp2kao -lcp2kinput -lcp2kpw -lcp2kgpu -lcp2kfft -lcp2kcommon -lcp2koffload -lcp2kmpiwrap -lcp2kfpga -lcp2kbase -L/var/folders/q_/hmjtsjld2d7csxq_lt8ttrtc0000gn/T/tiziano/spack-stage/spack-stage-cp2k-2023.1-6mcuxgwfs2ssvqffrkra7stb3hjzfacu/spack-src/lib/darwin-ventura-skylake-gcc/psmp/exts/dbcsr -ldbcsr /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_intel_lp64.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_sequential.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_core.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_intel_lp64.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_sequential.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_core.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_intel_lp64.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_sequential.dylib /opt/intel/oneapi/mkl/2023.0.0/lib/libmkl_core.dylib /Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/netlib-scalapack-2.2.0-ymqt25ppsjrbpxn2pg34fbwvrkg4osml/lib/libscalapack.dylib /usr/local/lib/libmpi.dylib -lstdc++ -L/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/libint-2.6.0-zhcvkhf3zzyociu7ep22oxzqak2koiws/lib -lint2 -L/Users/tiziano/Projects/spack/opt/spack/darwin-ventura-skylake/gcc-12.2.0/libxc-6.1.0-ipsgzeujk2uhcamoeglwxgdw62dyc3fb/lib -lxcf03 -lxc
...
```

What you see here is that `intel-oneapi-mkl` is in the dependency tree, yet `-D__MKL` is not present.

Comment (dev-zero) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: And here is the interactive demo:

```
>>> '^openmpi' in c
True
>>> '^mpich' in c
False
>>> '^mpi' in c
True
>>> '^intel-oneapi-mkl' in c
False
>>> c['blas'].name
'intel-oneapi-mkl'
>>> 'intel-oneapi-mkl' in c
True
```

Comment (alalazo) [reply var/spack/repos/builtin/packages/cp2k/package.py:277]: @dev-zero I could reproduce the issue trying to concretize `cp2k ^intel-oneapi-mkl` thanks. I'll open a new Issue on Github to triage that (which seems a serious bug)

**Review by alalazo — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@spackbot run pipeline

**Pull request merged**